zebra-chain/src/transaction.rs

```diff
@@ -318,14 +318,39 @@ impl Transaction {
     }
 
     /// Get this transaction's lock time.
-    pub fn lock_time(&self) -> LockTime {
-        match self {
-            Transaction::V1 { lock_time, .. } => *lock_time,
-            Transaction::V2 { lock_time, .. } => *lock_time,
-            Transaction::V3 { lock_time, .. } => *lock_time,
-            Transaction::V4 { lock_time, .. } => *lock_time,
-            Transaction::V5 { lock_time, .. } => *lock_time,
+    pub fn lock_time(&self) -> Option<LockTime> {
+        let lock_time = match self {
+            Transaction::V1 { lock_time, .. }
+            | Transaction::V2 { lock_time, .. }
+            | Transaction::V3 { lock_time, .. }
+            | Transaction::V4 { lock_time, .. }
+            | Transaction::V5 { lock_time, .. } => *lock_time,
+        };
+
+        // A lock time with a block height of zero means in practice that the lock time is
+        // disabled because it was effectively unlocked at the genesis block.
+        if lock_time == LockTime::unlocked() {
+            return None;
         }
+
+        // Consensus rule:
+        //
+        // > The transaction must be finalized: either its locktime must be in the past (or less
+        // > than or equal to the current block height), or all of its sequence numbers must be
+        // > 0xffffffff.
+        //
+        // This means that the lock time is only enabled if at least one input sequence number is
+        // not `u32::MAX`. The code below will make the method return `None` if such sequence
+        // number isn't found.
+        //
+        // https://developer.bitcoin.org/devguide/transactions.html#non-standard-transactions
+        let _sequence_number_enabling_lock_time = self
+            .inputs()
+            .iter()
+            .map(transparent::Input::sequence)
+            .find(|sequence_number| *sequence_number != u32::MAX)?;
+
+        Some(lock_time)
```

Comment:
Optional Refactor: This code seems more explicit.

I also moved the comment that explains what the code does, so it's right next to the code.

```suggestion

        // Only return the lock time if a transparent sequence number enables it.
        if self
            .inputs()
            .iter()
            .map(transparent::Input::sequence)
            .any(|sequence_number| *sequence_number != u32::MAX) {
                Some(lock_time)
        } else {
                None
        }

```

Reply:
Done in [8da0661](https://github.com/jvff/zebra/commit/8da06612d22319e32bf23825696398f567638a61), but I didn't add the extra comment because I thought that the binding name would be enough, but I can add it back if you think it's better.